test/Test/MockApi.hs

```diff
@@ -0,0 +1,227 @@
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+-- | A framework for Mocking API effects for testing
+--
+-- This system resolves around setting expectations.  An expectation represents a
+-- request we expect to be made and the result of that request.  Additionally, the
+-- expectation may be that the request is made only once or many times.
+--
+-- The request must be a @FossaApiClientF@ and must exactly be the request
+-- expected.  This is due to some limitations on matching types at run time.
+-- Expectations are matched in the order they are set up.  An assertion failure
+-- will be raised if a request is made that doesn't have a matching expectation.
+--
+-- The result can either be a return value or to die with a fatal diagnostic
+-- error.
+--
+-- Set up expectations using the `returnsOnce`, `alwaysReturns` and `fails`
+-- helpers.
+--
+-- @
+--     -- Set up basic expectation
+--     GetOrganization `alwaysReturns` Fixtures.organization
+--     GetProject revision `returnsOnce` Fixtures.project
+--
+--     -- Expect get issues to be called and fail, then return successfully when
+--     -- called again.
+--     GetIssues revision `fails` "Mock API error"
+--     GetIssues revision `returnsOnce` Fixtures.issues
+-- @
+--
+-- An expectation is considered satisfied if it had a limited number of invocations
+-- and it has been matched that many times.  If any expectations are unsatisfied at
+-- the end of the test, an test assertion failure will be raised.
+module Test.MockApi (
+  ApiExpectation,
+  FossaApiClientMockC,
+  MockApi (..),
+  MockApiC (runMockApiC),
+  alwaysReturns,
+  assertAllSatisfied,
+  fails,
+  returnsOnce,
+  runMockApi,
+  runApiWithMock,
+) where
+
+import Control.Algebra (Algebra (..), Has, send, type (:+:) (..))
+import Control.Carrier.Simple (SimpleC, interpret)
+import Control.Carrier.State.Strict (StateC (StateC), runState)
+import Control.Effect.Diagnostics (Diagnostics, fatalText)
+import Control.Effect.FossaApiClient (FossaApiClientF (..))
+import Control.Effect.Lift (Lift, sendIO)
+import Control.Effect.State (State, get, modify, put)
+import Control.Monad (guard)
+import Control.Monad.Trans (MonadIO)
+import Data.Kind (Type)
+import Data.List (intercalate)
+import Data.Text (Text)
+import Test.HUnit (assertFailure)
+
+data MockApi (m :: Type -> Type) a where
+  MockApiOnce :: FossaApiClientF a -> a -> MockApi m ()
+  MockApiAlways :: FossaApiClientF a -> a -> MockApi m ()
+  MockApiFails :: FossaApiClientF a -> Text -> MockApi m ()
+  MockApiRunExpectations :: FossaApiClientF a -> MockApi m (Maybe (ApiResult a))
+  AssertUnexpectedCall :: FossaApiClientF a -> MockApi m a
+  AssertAllSatisfied :: MockApi m ()
+
+-- | Specifies how to handle repetition of a request
+data ExpectationRepetition = Once | Always deriving (Eq, Ord, Show)
+
+-- | The result of a call can be either a value or a diagnostic failure
+data ApiResult a = Return a | Die Text deriving (Eq, Ord, Show)
```

Comment:
I did this because I was originally thinking there might be more cases, but these were the only two I had so far.  I could see some more specific ones like timeouts, but it seems most of our errors are communicated via diagnostic messages and not by using another exception handling effect.

I sort of like your second example there.  Let me try that.  Also, fortunately, `Either Text` is a pretty common way to express errors, so that's easy enough to read.

Reply:
Did this.  I'm not sure it reads better because pattern matching becomes wordy like `ApiResult (Left (ApiFail msg))` instead of just `Die msg`. I'm not too particular though.

It would be cool if `Either` were a typeclass that could be derived instead.